libsmtutil/Z3CHCInterface.cpp

```diff
@@ -82,9 +88,18 @@ pair<CheckResult, CHCSolverInterface::CexGraph> Z3CHCInterface::query(Expression
 		case z3::check_result::sat:
 		{
 			result = CheckResult::SATISFIABLE;
-			auto proof = m_solver.get_answer();
-			auto cex = cexGraph(proof);
-			return {result, cex};
+			// z3 version 4.8.8 modified Spacer to also return
+			// proofs containing nonlinear clauses.
```

Comment:
Nicer than checking the version would be to find out if Spacer actually gives us what we need by examining the proof expression... but since the segfaults were within ``Spacer``, resp. Z3 itself I guess that's not so easy? If not then a version check (but with a proper version compare ;-)) is fine...

Reply:
True, let me check

Reply:
Ideally ``cexGraph`` could just run on any version and optimistically hope that it'll get something usable, but be able to notice, if it doesn't and abort before Z3 has a chance to segfault...
Who knows if the proof structure won't change in the future and especially for the dlopen stuff that wouldn't be nice...
But I'm not quite clear where the segfault was coming from, so maybe it's infeasible.

Reply:
Yea, I'm still recompiling with 4.8.7, takes forever... I wanna check where exactly the seg fault is. If it's inside `get_proof` there's nothing we can do besides the version. Otherwise we might be able to do what you said

Reply:
```
#19 0x0000555555ec326c in z3::fixedpoint::get_answer (this=0x5555564884c0) at /usr/local/include/z3++.h:2862
#20 0x0000555555ec15dc in solidity::smtutil::Z3CHCInterface::query (this=0x5555564884a0, _expr=...) at solidity/libsmtutil/Z3CHCInterface.cpp:101
```
The crash happens inside `get_answer`, I suspect because their proof code didn't like nonlinear systems then, which is exactly what changed in 4.8.8. I'm afraid we have to go with the version solution.